### PR TITLE
New lint: `iter_count`

### PR DESCRIPTION
**Pull request by TaKO8Ki**
This pull request adds a new lint named `iter_count`.

---

closes https://github.com/rust-lang/rust-clippy/issues/6262

changelog: new lint `iter_count`

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
Should this also lint for `into_iter().count()` ?

**Comment by TaKO8Ki** [comment]
@matthewjasper 
Oh, I forgot to do this. I'm going to fix the lint.

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -2588,6 +2617,39 @@ fn lint_iter_next<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>, iter_
     }
 }
 
+fn lint_iter_count<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>, iter_args: &'tcx [Expr<'tcx>], is_mut: bool) {
```
Comment (Y-Nak) [inline clippy_lints/src/methods/mod.rs:2620]: Could you move the function to its own file (maybe `iter_count.rs`) as `bytes_nth` does in order not to make `methods/mod.rs` messy anymore? Please see #6680 for details.

Comment (TaKO8Ki) [reply clippy_lints/src/methods/mod.rs]: I've just moved the function to `iter_count.rs`!

**Comment by TaKO8Ki** [comment]
@flip1995 @matthewjasper @Y-Nak 
I'm ready for your review!

**Review by matthiaskrgr — CHANGES_REQUESTED** [review]
Seems there is a false positive with code inside the cargo repo.
Reduced:


needs a cargo.toml with
````
[dependencies]
im-rc = "*"
````

main.rs:
````rust
fn main() {
    let mut g: Graph<i32, i32> = Graph::new();
    g.iter().count();
}

impl<N: Eq + Ord + Clone, E: Default + Clone> Graph<N, E> {
    pub fn new() -> Graph<N, E> {
        Graph {
            nodes: im_rc::OrdMap::new(),
        }
    }

    pub fn iter(&self) -> impl Iterator<Item = &N> {
        self.nodes.keys()
    }
}

pub struct Graph<N: Clone, E: Clone> {
    nodes: im_rc::OrdMap<N, im_rc::OrdMap<N, E>>,
}
````

Your lint warns 
````
warning: called `.iter().count()` on a `std::iter::Iterator`
 --> src/main.rs:3:5
  |
3 |     g.iter().count();
  |     ^^^^^^^^^^^^^^^^ help: try: `g.len()`
````

but there is no corresponding `.len()` method

````
error[E0599]: no method named `len` found for struct `Graph<i32, i32>` in the current scope
  --> src/main.rs:3:7
   |
3  |     g.len();
   |       ^^^ method not found in `Graph<i32, i32>`
...
18 | pub struct Graph<N: Clone, E: Clone> {
   | ------------------------------------ method `len` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `len`, perhaps you need to implement one of them:
           candidate #1: `std::iter::ExactSizeIterator`
           candidate #2: `typenum::type_operators::Len`
           candidate #3: `bitmaps::types::BitOps`

````

**Comment by TaKO8Ki** [comment]
@matthewjasper 
I've just solved this false positive.

File: clippy_lints/src/methods/iter_count.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+pub(crate) fn lints<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>, iter_args: &'tcx [Expr<'tcx>], is_mut: bool) {
+    let mut_str = if is_mut { "_mut" } else { "" };
+    let iter_method = if method_chain_args(expr, &[format!("iter{}", mut_str).as_str(), "count"]).is_some() {
+        "iter"
+    } else if method_chain_args(expr, &["into_iter", "count"]).is_some() {
+        "into_iter"
+    } else {
+        return;
+    };
```
Comment (giraffate) [inline clippy_lints/src/methods/iter_count.rs:22]: Are these lines needed? I guess receiving a iter-method name as an arg seems to be enough, but what about?

File: clippy_utils/src/lib.rs [hunk 3]
```diff
@@ -1786,6 +1786,46 @@ pub fn is_some_ctor(cx: &LateContext<'_>, res: Res) -> bool {
     false
 }
 
+pub fn derefs_to_slice<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'tcx>,
+    ty: Ty<'tcx>,
+) -> Option<&'tcx hir::Expr<'tcx>> {
```
Comment (Y-Nak) [inline clippy_utils/src/lib.rs:1793]: Could you add documentation that explains this function?
Also, I'm not sure it's necessary to move this under `clippy_utils`.

File: clippy_lints/src/methods/iter_count.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    let ty = cx.typeck_results().expr_ty(&iter_args[0]);
+    if_chain! {
+        let caller_type = if derefs_to_slice(cx, &iter_args[0], ty).is_some() {
```
Comment (Y-Nak) [inline clippy_lints/src/methods/iter_count.rs:25]: This seems unnecessary `if_chain`.

Comment (TaKO8Ki) [reply clippy_lints/src/methods/iter_count.rs]: I removed it.

File: clippy_utils/src/lib.rs (hunk 3 above)
Comment (TaKO8Ki) [reply clippy_utils/src/lib.rs:1793]: Now [`lint_iter_cloned_collect`](https://github.com/rust-lang/rust-clippy/blob/d5223be2e30a9d116b839bda418bec99d2949a68/clippy_lints/src/methods/mod.rs#L2424), [`lint_iter_nth`](https://github.com/rust-lang/rust-clippy/blob/d5223be2e30a9d116b839bda418bec99d2949a68/clippy_lints/src/methods/mod.rs#L2603), [`lint_iter_next`](https://github.com/rust-lang/rust-clippy/blob/d5223be2e30a9d116b839bda418bec99d2949a68/clippy_lints/src/methods/mod.rs#L2546) and so on use `derefs_to_slice` defined in clippy_lints/src/methods/mod.rs. And I want to use it in iter_count.rs. That's why I  moved it under `clippy_utils`.
You mean that I should redefine `derefs_to_slice` in `iter_count.rs` or use the other function?

File: clippy_lints/src/methods/iter_count.rs (hunk 2 above)
Comment (TaKO8Ki) [reply clippy_lints/src/methods/iter_count.rs]: I've just modified `lints()` in order to receive iter method name as an argument.

File: clippy_utils/src/lib.rs (hunk 3 above)
Comment (Y-Nak) [reply clippy_utils/src/lib.rs:1793]: I think this function is not so general that can be used from the outside of `methods`.  You can use it in `iter_count.rs` without moving it to `clippy_utils`, right? Or you can add another module `methods/utils.rs` and move it under the module.

Comment (TaKO8Ki) [reply clippy_utils/src/lib.rs]: Thank you. I moved `derefs_to_slice` to `methods/mod.rs` and made it visible within the crate in order to use it in `iter_count.rs`.

Comment (Y-Nak) [reply clippy_utils/src/lib.rs]: You don't need to add `pub(crate)` to use it inside `methods` module.

Comment (TaKO8Ki) [reply clippy_utils/src/lib.rs]: Oh, I was mistaken thinking that `derefs_to_slice` and `iter_count.rs` are in different modules.

File: clippy_lints/src/methods/iter_count.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        "BTreeMap"
+    } else if match_type(cx, ty, &paths::BTREESET) {
+        "BTreeSet"
+    } else if match_type(cx, ty, &paths::LINKED_LIST) {
+        "LinkedList"
+    } else if match_type(cx, ty, &paths::BINARY_HEAP) {
+        "BinaryHeap"
+    } else {
+        return;
+    };
```
Comment (matthiaskrgr) [inline clippy_lints/src/methods/iter_count.rs:33]: I'm wondering if there is a more elegant way to do this, hm.

Comment (TaKO8Ki) [reply clippy_lints/src/methods/iter_count.rs:33]: I think so, too.
I'm going to find a better way.

Comment (TaKO8Ki) [reply clippy_lints/src/methods/iter_count.rs:33]: Hmm. `ty.to_string()` show more detailed information than I need.

File: clippy_lints/src/methods/iter_count.rs (hunk 2 above)
Comment (giraffate) [reply clippy_lints/src/methods/iter_count.rs]: It looks good. Thanks!

File: clippy_lints/src/methods/iter_count.rs (hunk 5 above)
Comment (giraffate) [reply clippy_lints/src/methods/iter_count.rs:33]: If there was a better way, it's so good. But I guess this isn't a blocker to approve.

**Review by matthiaskrgr — APPROVED** [review]
Thanks!

**Pull request merged**